oaknational/oak-ai-lesson-assistant
# test: seeding a pool of test users for e2e tests

## Problem

At the moment, all our e2e tests use a single test@thenational.academy user. This is limiting us because:

- That account isn't a demo user, so we can't test demo functionality
- The account is reused between tests, so we can't test rate limits, user bans, other stateful behaviour
- The single clerk account has different database state on each dev's machine

## Description

- `test-support` tRPC router and `prepareUser` mutation
  - Prepares a test user for the upcoming test
    - Uses an email address pattern to reuse a pool of accounts
      - Test users are unique by variant. A variant will set up the account in a certain state
      - Test users are unique by environment. An environment can be a vercel preview branch or a local machine hostname
      - eg: `test+my-branch-demo@thenational.academy`, `test+matts-macbook-demo@thenational.academy`
    - Resets and returns the matching clerk user. Or creates it if missing
    - Returns a single use sign-in token
- `test-support` nextjs page
  - Creates a clerk session with a sign-in token 
- demo e2e test
  - An example of using the new system to test an account in the demo state

Not included at the moment:
- Preparing users and caching their cookies in a setup step, like our current auth pattern. This would save us a few seconds for each test but adds complexity we don't need yet

## Issue(s)

Fixes #AI-553

## How to test

1. run `pnpm run test-e2e-ui`
2. Navigate to `demo-accounts.test.ts` and the test inside
3. Press play
4. Check [clerk](https://dashboard.clerk.com/apps/app_2U3lg9YwuCTyYRgNWcN3jxIeTGV/instances/ins_2U3lgGV1Nc6kBEvlEgZz7271OKk/users) to see the new user created



## Screenshots

First run:
![CleanShot 2024-09-25 at 11 48 42@2x](https://github.com/user-attachments/assets/2d020334-ad1e-4f9f-adef-8d0800d411a1)

Second run:
![CleanShot 2024-09-25 at 11 49 00@2x](https://github.com/user-attachments/assets/51389b1a-8ae7-4863-a3bd-7057b9e369ba)

## Checklist
